assets/js/_helpers.js

```diff
@@ -0,0 +1,249 @@
+                    else
+                        callbacks.on200(xhr.response);
+                }
+            } else {
+                // handled by onerror
+                if (xhr.status === 0) return;
+
+                if (callbacks.onNon200)
+                    callbacks.onNon200(xhr);
+            }
+        };
+
+        xhr.ontimeout = function () {
+            if (callbacks.onTimeout)
+                callbacks.onTimeout(xhr);
+        };
+
+        xhr.onerror = function () {
+            if (callbacks.onError)
+                callbacks.onError(xhr);
+        };
+
+        if (options.payload)
+            xhr.send(options.payload);
+        else
+            xhr.send();
+    },
+    /** @private */
+    _xhrRetry: function(method, url, options, callbacks) {
+        if (options.retries <= 0) {
+            console.warn('Failed to pull', options.entity_name);
+            if (callbacks.onTotalFail)
+                callbacks.onTotalFail();
+            return;
+        }
+        helpers._xhr(method, url, options, callbacks);
+    },
+    /**
+     * @callback callbackXhrOn200
+     * @param {Object} response - xhr.response
+     */
+    /**
+     * @callback callbackXhrError
+     * @param {XMLHttpRequest} xhr
+     */
+    /**
+     * @param {'GET'|'POST'} method - 'GET' or 'POST'
+     * @param {String} url - URL to send request to
+     * @param {Object} options - other XHR options
+     * @param {XMLHttpRequestBodyInit} [options.payload=null] - payload for POST-requests
+     * @param {'arraybuffer'|'blob'|'document'|'json'|'text'} [options.responseType=json]
+     * @param {Number} [options.timeout=10000]
+     * @param {Number} [options.retries=1]
+     * @param {String} [options.entity_name='unknown'] - string to log
+     * @param {Number} [options.retry_timeout=1000]
+     * @param {Object} callbacks - functions to execute on events fired
+     * @param {callbackXhrOn200} [callbacks.on200]
+     * @param {callbackXhrError} [callbacks.onNon200]
+     * @param {callbackXhrError} [callbacks.onTimeout]
+     * @param {callbackXhrError} [callbacks.onError]
+     * @param {callbackXhrError} [callbacks.onTotalFail] - if failed after all retries
+     */
+     xhr: function(method, url, options, callbacks) {
+        if (!options.retries || options.retries <= 1) {
+            helpers._xhr(method, url, options, callbacks);
+            return;
+        }
+
+        if (!options.entity_name) options.entity_name = 'unknown';
+        if (!options.retry_timeout) options.retry_timeout = 1000;
+        const retries_total = options.retries;
+        let currentTry = 1;
+
+        const retry = function () {
+            console.warn('Pulling ' + options.entity_name + ' failed... ' + (currentTry++) + '/' + retries_total);
+            setTimeout(function () {
+                options.retries--;
+                helpers._xhrRetry(method, url, options, callbacks);
+            }, options.retry_timeout);
+        };
+        
```

Comment:
extra whitespaces here

Reply:
[The root of trouble resolved](https://stackoverflow.com/a/30884298) :)